### PR TITLE
[WIP] Add proper annotation metadata

### LINKED ISSUE #115: Make annotation metadata actual metadata
Annotations currently support the `{ :msg "some label" }` syntax. The intent of the syntax is to support arbitrary metadata tags there, however currently there is only the `:msg` tag *hardcoded* in the grammar. In anticipation of allowing other tags (e.g. we are discussing a `skipConstructor` boolean metadata tag), we should:

1. Fix the grammar and data-types to support a comma separated list of `:tag value`

2. Check that the values provided correspond to the expected type of the tag (so for exampel ":msg 1" should be an error)

### PR DESCRIPTION
**Pull request by cd1m0**
NOTE: This PR depends on #112 . After that merges this should be re-targeted to develop.

This PR doesn't have many visible changes. It tweaks the grammar and the datatypes inside `SAnnotation` to allow adding more metadata tags other than just `:msg`. Closes #115

### CODE REVIEW

**Review by blitz-1306 — APPROVED** [review]

File: src/spec-lang/expr_grammar.pegjs [hunk 1]
```diff
@@ -29,15 +29,48 @@ AnnotationStr =
...
+                acc[el[0]] = el[1]
+                return acc;
+            },
+            base
+        );
+    }
+
+
+AnnotationMD =
+    "{" __  exprs: MDExpressionList __ "}" { return exprs; }
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/spec-lang/expr_grammar.pegjs]: Fixed in 9156843.

File: src/spec-lang/ast/declarations/annotation.ts [hunk 2]
```diff
@@ -10,15 +13,42 @@ export enum AnnotationType {
...
+        for (const [key, val] of Object.entries(this.md)) {
+            const expectedType = knownMDTypes.get(key);
+
+            if (!expectedType) {
+                throw new Error(`Unknown annotation metadata key ${key}`);
+            }
+
+            if (!(val instanceof expectedType)) {
+                throw new Error(
+                    `Expected key ${key} to be a ${expectedType}, not ${val} of type ${typeof val}`
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/spec-lang/ast/declarations/annotation.ts]: Fixed in 9156843.

File: src/spec-lang/expr_grammar.pegjs [hunk 3]
```diff
@@ -110,7 +143,7 @@ IndexPath =
 // TODO: Eventually remove hacky '/' from if_updated rule. This is to work around
 // limitations in Solidity - it throws if it sees natspec on internal state vars
 If_Updated =
-    ("/" __)? type: IF_UPDATED __ label: AnnotationLabel? __ expr: Expression __ ";" {
+    ("/" __)? type: IF_UPDATED __ label: AnnotationMD? __ expr: Expression __ ";" {
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/spec-lang/expr_grammar.pegjs]: Fixed in 9156843.

File: src/spec-lang/expr_grammar.pegjs [hunk 4]
```diff
@@ -120,7 +153,7 @@ If_Updated =
     }
 
 If_Assigned =
-    ("/" __)? type: IF_ASSIGNED path: IndexPath __ label: AnnotationLabel? __ expr: Expression __ ";" {
+    ("/" __)? type: IF_ASSIGNED path: IndexPath __ label: AnnotationMD? __ expr: Expression __ ";" {
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/spec-lang/expr_grammar.pegjs]: Fixed in 9156843.

File: src/spec-lang/expr_grammar.pegjs [hunk 5]
```diff
@@ -151,7 +184,7 @@ TypedArgs =
     }
 
 UserFunctionDefinition =
-    type: DEFINE __ label: AnnotationLabel? __ name: Identifier __ "(" __ args: TypedArgs? __ ")" __ returnType: Type __ "=" __ body: Expression {
+    type: DEFINE __ label: AnnotationMD? __ name: Identifier __ "(" __ args: TypedArgs? __ ")" __ returnType: Type __ "=" __ body: Expression {
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/spec-lang/expr_grammar.pegjs]: Fixed in 9156843.

**Review by blitz-1306 — APPROVED** [review]

**Comment by cd1m0** [comment]
Intermitten segfaults in CI again. Force merging

**Pull request merged**